hibernate-rx-core/src/main/java/org/hibernate/rx/persister/entity/impl/HiLoRxOptimizer.java

```diff
@@ -0,0 +1,56 @@
+package org.hibernate.rx.persister.entity.impl;
+
+import org.hibernate.AssertionFailure;
+import org.hibernate.id.IdentifierGeneratorHelper;
+import org.hibernate.id.enhanced.HiLoOptimizer;
+
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+public class HiLoRxOptimizer extends HiLoOptimizer implements RxOptimizer<Long> {
+
+	public HiLoRxOptimizer(Class<?> integerType, int incrementSize) {
+		super( integerType, incrementSize );
+	}
+
+	@Override
+	public synchronized CompletionStage<Optional<Long>> generate(RxAccessCallback callback) {
+
+		final GenerationState generationState = locateGenerationState( callback.getTenantIdentifier() );
+
+		if ( !generationState.isInitialized() ) {
+			return getValidFirstValueFromDatabase( callback )
+					.thenCompose( validFirstValue -> getActualNextValue( generationState, validFirstValue ) );
+		}
+		else if ( generationState.requiresNewSourceValue() ) {
+			return callback.getNextValue()
+					.thenCompose( valueFromDatabase -> getActualNextValue( generationState, valueFromDatabase ) );
+		}
+		else {
+			return CompletableFuture.completedFuture(Optional.of(generationState.makeValueThenIncrement().longValue()));
```

Comment:
When `GenerationState` is in such a state that clients needing a new ID reach this line, they will all be creating a `CompletableFuture` wrapping `makeValueThenIncrement`: so you risk having thousands of threads hitting the DB to allocate a new block.

In Hibernate ORM this is prevented by the `synchronized` keyword: all threads need to wait for the first increment to have completed; in this case we're by-passing the lock, which is going to flood the DB with increase requests - that's perhaps "correct" but it's side-stepping the point of the optimiser :-/

I'm afraid this needs to be modelled in a different way. If there's one operation of `makeValueThenIncrement` being scheduled or in progress, the other operations need to wait on that same operation, and then check again for the state as it's possible that a single increment operation didn't allocate enough IDs for all clients.

Reply:
@Sanne is is really the case that you're likely to have thousands of inserts happening simultaneously?

I mean, a system which does 1000 inserts a second is adding almost 100 million rows to the table each day. That can't possibly be the common case can it?

Reply:
Fair enough, it might be a bit pathological - but that's why I'm saying that I'd rather see a "dumb" optimiser implemented first, as the whole point of the current HiLoOptimiser is to optimise for such pathological cases?

But also consider, that a system that _happens to_ issue 1000 inserts in a second in a specific point in time, might be having a momentary spike; it's totally possible that there won't be any other insert for the rest of the day - so while doing 100 million inserts a day might be a stretch, we still need to handle spikes graciously.  I'd argue such spikes are not far fetched, there might be the "daily upload" of data and processes like that - just saying let's not look at average expected rates. (But if we did, 100M inserts a day are not unlikely either!)

Consider what happens when the in-memory pool is exhausted - which will happen periodically and quite often - and you have a non trivial amount of operations "on the stack" of the various threads running this RX system (which supposedly is good for highly concurrent scenarios): you'll have _ALL_ operations that hit this method also hit the database, for quite a long span of time:

the T1 starting point of time is when we first exhausted the pool
the T2 end point is when the first completable stage actually returned from the DB (and replenished the pool)

_ALL_ operations between T2-T1 will have caused an increse in the "page" on the DB. So let's focus on that a second: we're not skipping individual IDs, but whole pages of {sizeIncrement}, multipled by the amount of parallel ops that this reactive system is able to take... Worst aspect is that soon after, we'll repeat this again: I'm afraid we'll exhaust any sequence rather fast.

In terms of design, I believe the goal of such optimisers is to prevent such a picture - however I can agree that this is not a level of optimisations we need for a first MVP of the project: just explaining why I believe current Optimisers can't be taken as-is.

Reply:
> But also consider, that a system that happens to issue 1000 inserts in a second in a specific point in time, might be having a momentary spike; it's totally possible that there won't be any other insert for the rest of the day

Sure sure, I know that's possible. I'm just saying that even to get a spike like that I think you would have to be dealing with something pretty specific, and it's reasonable that such general purpose optimizers might not be the right thing to use for something so specific.

I mean, sure, you *can* get such an avalanche by chance, but only with [extremely low probability](https://en.wikipedia.org/wiki/Poisson_distribution).

> I'd argue such spikes are not far fetched, there might be the "daily upload" of data and processes like that

Yes of course, but a "daily upload" is more likely to happen in a single thread, not distributed across 1000 threads.

> In terms of design, I believe the goal of such optimisers is to prevent such a picture - however I can agree that this is not a level of optimisations we need for a first MVP of the project: just explaining why I believe current Optimisers can't be taken as-is.

Look, TBH I don't have a strong feeling about how optimizers should work *at all*. I would just like it if we made some sort of basic no-frills best effort to respect the `allocationSize` our users have explicitly specified in `@TableGenerator` or `@SequenceGenerator`, so we can cross that off the list of Limitations, i.e. of unsupported features of JPA.

Sure, sure, our initial implementation might not exhibit ideal behavior for some particularly hardcore workloads, but at least for the most common case where inserts arrive in a steady flow, according to a typical Poisson distribution, rather than in sudden avalanches, it would cut the wait time in half, by avoiding an additional roundtrip on every insert.

(I don't think we're really disagreeing over very much here.)

Reply:
right we're not disagreeing: let's forget about Hi/Lo optimisers for now and focus on the no-frills implementations we really need.

:+1: